src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs

```diff
@@ -0,0 +1,77 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    class NoRefOrLibFolderInPackageRule : IPackageRule
+    {
+        private static readonly string BuildDir = PackagingConstants.Folders.Build + Path.DirectorySeparatorChar;
+        public string MessageFormat { get; }
+
+        public NoRefOrLibFolderInPackageRule(string messageFormat)
+        {
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader package)
+        {
+            var files = package.GetFiles();
+            return Validate(files);
+        }
+
+        internal IEnumerable<PackagingLogMessage> Validate(IEnumerable<string> files)
+        {
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var collection = new ContentItemCollection();
+            collection.Load(files);
+
+            var libItems = GetContentForPattern(collection, managedCodeConventions.Patterns.CompileLibAssemblies);
+            var refItems = GetContentForPattern(collection, managedCodeConventions.Patterns.CompileRefAssemblies);
+            var buildItems = GetContentForPattern(collection, managedCodeConventions.Patterns.MSBuildFiles);
+
+            var libFrameworks = GetGroupFrameworks(libItems).ToArray();
+            var refFrameworks = GetGroupFrameworks(refItems).ToArray();
+            var buildFrameworks = GetGroupFrameworks(buildItems).ToArray();
+
+            if (libFrameworks.Length == 0 && refFrameworks.Length == 0)
+            {
+                //if you can't find the ref and lib folder, then find the build folder
+                if (buildFrameworks.Length != 0)
+                {
+                    //if you can find any folders other than native or any, raise an NU5127
+                    if (buildFrameworks.
```

Comment:
If you use the NuGetFramework object you can check something like `IsSpecificFramework`
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Frameworks/NuGetFramework.cs#L324